File: src/world/ConstructionVocabulary.js
/**
 * Canonical MVP construction vocabulary and renderer-independent catalog
 * disclosure policy. Stable building IDs remain in BuildingCatalog; this file
 * owns the smaller set of concepts shared by content, saves, editor rules, and
 * presentation.
 */

export const MVP_ZONE_IDS = Object.freeze({
  RESIDENTIAL: 'RESIDENTIAL',
  COMMERCIAL: 'COMMERCIAL',
  OPERATIONS: 'OPERATIONS'
});

export const MVP_ZONE_LABELS = Object.freeze({
  [MVP_ZONE_IDS.RESIDENTIAL]: 'Residential',
  [MVP_ZONE_IDS.COMMERCIAL]: 'Commercial',
  [MVP_ZONE_IDS.OPERATIONS]: 'Operations'
});

export const CONSTRUCTION_CATEGORIES = Object.freeze({
  RESIDENTIAL: 'Residential',
  COMMERCIAL: 'Commercial',
  OPERATIONS: 'Operations',
  FACILITIES: 'Facilities',
  INFRASTRUCTURE: 'Infrastructure'
});

export const CATALOG_STAGES = Object.freeze({
  STARTER: 'STARTER',
  ADVANCED: 'ADVANCED'
});

export const PROGRESSION_TIERS = Object.freeze({
  OPERATOR: 'OPERATOR',
  BROKER: 'BROKER',
  MAGNATE: 'MAGNATE'
});

const TIER_RANK = Object.freeze({
  [PROGRESSION_TIERS.OPERATOR]: 1,
  [PROGRESSION_TIERS.BROKER]: 2,
  [PROGRESSION_TIERS.MAGNATE]: 3
});

const ZONE_ALIASES = Object.freeze({
  RES: MVP_ZONE_IDS.RESIDENTIAL,
  RESIDENTIAL: MVP_ZONE_IDS.RESIDENTIAL,
  COM: MVP_ZONE_IDS.COMMERCIAL,
  COMMERCIAL: MVP_ZONE_IDS.COMMERCIAL,
  OFFICE: MVP_ZONE_IDS.COMMERCIAL,
  OPS: MVP_ZONE_IDS.OPERATIONS,
  IND: MVP_ZONE_IDS.OPERATIONS,
  INDUSTRIAL: MVP_ZONE_IDS.OPERATIONS,
  OPERATIONS: MVP_ZONE_IDS.OPERATIONS,
  // These IDs are load-only compatibility records. New players construct
  // facilities instead of painting service zones.
  POWER: 'POWER_SERVICE',
  POWER_SERVICE: 'POWER_SERVICE',
  WATER: 'WATER_SERVICE',
  WATER_SERVICE: 'WATER_SERVICE',
  FIRE: 'FIRE_SERVICE',
  FIRE_SERVICE: 'FIRE_SERVICE',
  SUBURBAN_RESIDENTIAL: 'SUBURBAN_RESIDENTIAL'
});

export function normalizeZoneId(value) {
  if (typeof value !== 'string') return null;
  return ZONE_ALIASES[value.trim().toUpperCase()] || null;
}

export function isMvpDevelopmentZone(value) {
  const normalized = normalizeZoneId(value);
  return Object.values(MVP_ZONE_IDS).includes(normalized);
}

export function getZoneLabel(value) {
  const normalized = normalizeZoneId(value);
  return MVP_ZONE_LABELS[normalized] || normalized;
}

export function getUnlockedProgressionTier(values = {}) {
  let rank = TIER_RANK[PROGRESSION_TIERS.OPERATOR];
  for (const [tier, tierRank] of Object.entries(TIER_RANK)) {
    if (values?.[tier] === true) rank = Math.max(rank, tierRank);
  }
  return Object.keys(TIER_RANK).find(tier => TIER_RANK[tier] === rank)
    || PROGRESSION_TIERS.OPERATOR;
}

export function getCatalogAccess(spec, progression = {}) {
  if (!spec || typeof spec !== 'object') {
    throw new TypeError('catalog spec must be an object');
  }
  const requiredTier = spec.progressionTier || PROGRESSION_TIERS.OPERATOR;
  const currentTier = getUnlockedProgressionTier(progression);
  const requiredRank = TIER_RANK[requiredTier];
  if (!requiredRank) throw new RangeError(`Unknown catalog progression tier: ${requiredTier}`);
  const unlocked = TIER_RANK[currentTier] >= requiredRank;
  return Object.freeze({
    unlocked,
    currentTier,
    requiredTier,
    stage: spec.catalogStage || CATALOG_STAGES.ADVANCED,
    reason: unlocked ? null : `Unlocks at ${requiredTier[0]}${requiredTier.slice(1).toLowerCase()} tier`
  });
}

export function isCatalogSpecDisclosed(spec, { includeAdvanced = false } = {}) {
  return spec?.catalogStage === CATALOG_STAGES.STARTER || includeAdvanced;
}

File: godot/MetroPulse.Domain/Content/ConstructionVocabulary.cs
using System.Collections.ObjectModel;

namespace MetroPulse.Domain.Content;

public static class ConstructionCategories
{
    public const string Residential = "RESIDENTIAL";
    public const string Commercial = "COMMERCIAL";
    public const string Operations = "OPERATIONS";
    public const string Facilities = "FACILITIES";
    public const string Infrastructure = "INFRASTRUCTURE";

    public static readonly IReadOnlyList<string> All = Array.AsReadOnly([
        Residential,
        Commercial,
        Operations,
        Facilities,
        Infrastructure,
    ]);
}

public static class CatalogStages
{
    public const string Starter = "STARTER";
    public const string Advanced = "ADVANCED";
}

public static class ProgressionTiers
{
    public const string Operator = "OPERATOR";
    public const string Broker = "BROKER";
    public const string Magnate = "MAGNATE";
}

public sealed record CatalogAccess(
    bool Unlocked,
    string CurrentTier,
    string RequiredTier,
    string Stage,
    string? Reason);

/// <summary>
/// Exact renderer-independent construction vocabulary and catalog disclosure policy.
/// The UI may filter this projection, but it cannot grant access independently.
/// </summary>
public static class ConstructionVocabulary
{
    private static readonly IReadOnlyDictionary<string, string> ZoneAliases =
        new ReadOnlyDictionary<string, string>(new Dictionary<string, string>(StringComparer.Ordinal)
        {
            ["RES"] = ConstructionCategories.Residential,
            ["RESIDENTIAL"] = ConstructionCategories.Residential,
            ["COM"] = ConstructionCategories.Commercial,
            ["COMMERCIAL"] = ConstructionCategories.Commercial,
            ["OFFICE"] = ConstructionCategories.Commercial,
            ["OPS"] = ConstructionCategories.Operations,
            ["IND"] = ConstructionCategories.Operations,
            ["INDUSTRIAL"] = ConstructionCategories.Operations,
            ["OPERATIONS"] = ConstructionCategories.Operations,
            ["POWER"] = "POWER_SERVICE",
            ["POWER_SERVICE"] = "POWER_SERVICE",
            ["WATER"] = "WATER_SERVICE",
            ["WATER_SERVICE"] = "WATER_SERVICE",
            ["FIRE"] = "FIRE_SERVICE",
            ["FIRE_SERVICE"] = "FIRE_SERVICE",
            ["SUBURBAN_RESIDENTIAL"] = "SUBURBAN_RESIDENTIAL",
        });

    private static readonly IReadOnlyDictionary<string, int> TierRanks =
        new ReadOnlyDictionary<string, int>(new Dictionary<string, int>(StringComparer.Ordinal)
        {
            [ProgressionTiers.Operator] = 1,
            [ProgressionTiers.Broker] = 2,
            [ProgressionTiers.Magnate] = 3,
        });

    public static string? NormalizeZoneId(string? value)
    {
        if (string.IsNullOrWhiteSpace(value)) return null;
        return ZoneAliases.GetValueOrDefault(value.Trim().ToUpperInvariant());
    }

    public static bool IsMvpDevelopmentZone(string? value) => NormalizeZoneId(value) is
        ConstructionCategories.Residential or ConstructionCategories.Commercial or ConstructionCategories.Operations;

    public static string GetUnlockedTier(IEnumerable<string>? unlockedTiers)
    {
        int rank = TierRanks[ProgressionTiers.Operator];
        foreach (string tier in unlockedTiers ?? Array.Empty<string>())
        {
            if (TierRanks.TryGetValue(tier, out int candidate)) rank = Math.Max(rank, candidate);
        }
        return TierRanks.Single(entry => entry.Value == rank).Key;
    }

    public static CatalogAccess GetCatalogAccess(BuildingDefinition spec, IEnumerable<string>? unlockedTiers = null)
    {
        ArgumentNullException.ThrowIfNull(spec);
        string requiredTier = spec.ProgressionTier ?? ProgressionTiers.Operator;
        if (!TierRanks.TryGetValue(requiredTier, out int requiredRank))
        {
            throw new ArgumentOutOfRangeException(nameof(spec), requiredTier, "Unknown catalog progression tier.");
        }
        string currentTier = GetUnlockedTier(unlockedTiers);
        bool unlocked = TierRanks[currentTier] >= requiredRank;
        return new CatalogAccess(
            unlocked,
            currentTier,
            requiredTier,
            spec.CatalogStage ?? CatalogStages.Advanced,
            unlocked ? null : $"Unlocks at {ToTitle(requiredTier)} tier");
    }

    public static bool IsDisclosed(BuildingDefinition spec, bool includeAdvanced = false)
    {
        ArgumentNullException.ThrowIfNull(spec);
        return spec.CatalogStage == CatalogStages.Starter || includeAdvanced;
    }

    public static IReadOnlyList<BuildingDefinition> FilterCatalog(
        IEnumerable<BuildingDefinition> records,
        string? category = null,
        bool includeAdvanced = false,
        IEnumerable<string>? unlockedTiers = null,
        bool includeLocked = true)
    {
        ArgumentNullException.ThrowIfNull(records);
        BuildingDefinition[] filtered = records.Where(spec =>
            (string.IsNullOrWhiteSpace(category)
                || string.Equals(category, "ALL", StringComparison.Ordinal)
                || string.Equals(spec.Category, category, StringComparison.Ordinal))
            && IsDisclosed(spec, includeAdvanced)
            && (includeLocked || GetCatalogAccess(spec, unlockedTiers).Unlocked))
            .ToArray();
        return Array.AsReadOnly(filtered);
    }

    private static string ToTitle(string value) =>
        string.Concat(value[0], value[1..].ToLowerInvariant());
}
